src/function/table/table_scan.cpp

```diff
@@ -47,350 +51,461 @@ static StorageIndex GetStorageIndex(TableCatalogEntry &table, const ColumnIndex
+			                             l_state.scan_state.local_state.batch_index);
+		}
+		return OperatorPartitionData(0);
+	}
+};
+
+static unique_ptr<LocalTableFunctionState> TableScanInitLocal(ExecutionContext &context, TableFunctionInitInput &input,
+                                                              GlobalTableFunctionState *g_state) {
+	auto &cast_g_state = g_state->Cast<TableScanGlobalState>();
+	return cast_g_state.InitLocalState(context, input);
 }
 
-//===--------------------------------------------------------------------===//
-// Index Scan
-//===--------------------------------------------------------------------===//
-struct IndexScanGlobalState : public GlobalTableFunctionState {
-	IndexScanGlobalState(const data_ptr_t row_id_data, const idx_t count)
-	    : row_ids(LogicalType::ROW_TYPE, row_id_data), row_ids_count(count), row_ids_offset(0) {
+unique_ptr<GlobalTableFunctionState> DuckTableScanInitGlobal(ClientContext &context, TableFunctionInitInput &input,
+                                                             DataTable &storage, const TableScanBindData &bind_data) {
+	auto g_state = make_uniq<DuckTableScanState>(context, input.bind_data.get());
+	storage.InitializeParallelScan(context, g_state->state);
+	if (!input.CanRemoveFilterColumns()) {
+		return std::move(g_state);
 	}
 
-	const Vector row_ids;
-	const idx_t row_ids_count;
-	idx_t row_ids_offset;
-	ColumnFetchState fetch_state;
-	TableScanState local_storage_state;
-	vector<StorageIndex> column_ids;
-	bool finished;
-};
-
-static unique_ptr<GlobalTableFunctionState> IndexScanInitGlobal(ClientContext &context, TableFunctionInitInput &input) {
-	auto &bind_data = input.bind_data->Cast<TableScanBindData>();
+	g_state->projection_ids = input.projection_ids;
+	const auto &columns = bind_data.table.GetColumns();
+	for (const auto &col_idx : input.column_indexes) {
+		if (col_idx.IsRowIdColumn()) {
+			g_state->scanned_types.emplace_back(LogicalType::ROW_TYPE);
+		} else {
+			g_state->scanned_types.push_back(columns.GetColumn(col_idx.ToLogical()).Type());
+		}
+	}
+	return std::move(g_state);
+}
 
-	data_ptr_t row_id_data = nullptr;
-	if (!bind_data.row_ids.empty()) {
-		row_id_data = (data_ptr_t)&bind_data.row_ids[0]; // NOLINT - this is not pretty
+unique_ptr<GlobalTableFunctionState> DuckIndexScanInitGlobal(ClientContext &context, TableFunctionInitInput &input,
+                                                             DataTable &storage, const TableScanBindData &bind_data,
+                                                             unsafe_vector<row_t> &row_ids) {
+	auto g_state = make_uniq<DuckIndexScanState>(context, input.bind_data.get());
+	if (!row_ids.empty()) {
+		std::sort(row_ids.begin(), row_ids.end());
+		g_state->row_ids = std::move(row_ids);
+		auto row_id_data = (data_ptr_t)&g_state->row_ids[0]; // NOLINT - this is not pretty
+		g_state->row_id_vector = make_uniq<Vector>(LogicalType::ROW_TYPE, row_id_data);
```

Comment:
Can we remove this vector and instead when we are slicing this directly slice the row id data pointer?